### PR TITLE
First attempt to support installing cross compilation targets

### PR DESCRIPTION
**Pull request by bkchr**
Okay, now I need cross compilation support :D

I began to add support for installing cross compilation targets, but at the moment it is dirty. 

The problem is, that some targets like "mips-unknown-linux-musl" just exist for the "rust-std" extension. Next problem is, that if I install the rust package I have to add "rust-std" as extra extension although it already will be installed by the rust package.

Here is my current default.nix:
```nix
with import <nixpkgs> {}; 
let 
  myrust = (rustChannels.stable.rust.override { extensions = [ "rust-src" "rust-std" ]; extCrossTargets = [ "mips-unknown-linux-musl" ]; });
in
stdenv.mkDerivation {
  name = "rust-dev";
  buildInputs = [ myrust ];
}
```

I'm open for all kind of improvements^^

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/nixpkgs-mozilla/24)

### CODE REVIEW

File: rust-overlay.nix [hunk 1]
```diff
@@ -85,11 +95,13 @@ let
   fromManifest = manifest: { stdenv, fetchurl, patchelf }:
     let pkgs = fromTOML (builtins.readFile (builtins.fetchurl manifest)); in
     with super.lib; flip mapAttrs pkgs.pkg (name: pkg:
-      super.makeOverridable ({extensions}:
+      super.makeOverridable ({extensions, crossTargets, extCrossTargets}:
         let
+          targets = [(hostTripleOf stdenv.system)] ++ crossTargets;
```
Comment (nbp) [inline rust-overlay.nix]: This should only be `[(hostTripleOf stdenv.system) "*"]`, otherwise this will add issues such as patching the packages with the correct interpreter on a different architecture.

I think `crossTargets` and `extCrossTargets` should just be one attribute named `targets`.

**Comment by steveeJ** [comment]
Does this PR also help with installing the `x86_64-unknown-linux-musl` target?

**Comment by nbp** [comment]
@steveeJ , Yes it should let you write some argument to install `rust` with extensions (`rust-std` and `rust-analysis`) for the `x86_64-unknown-linux-musl` target.

**Comment by bkchr** [comment]
Hmm, why you don't want to install the arm toolchain as extra while you work with the x86_64 toolchain?

As I already said, the problem is that you can not install rust for "mips-unknown-linux-musl", because this architecture is only available for the extensions. I don't know why that is the case, but this prevents us to install this architecture toolchain standalone. 

I will address your comments to the code when I find some time :)

**Comment by nbp** [comment]
> Hmm, why you don't want to install the arm toolchain as extra while you work with the x86_64 toolchain?

Because, unless you setup your kernel to spawn qemu to execute mips binaries, these binaries are not going to be of any use.  Installing rust for mips on x86_64 is just useless as you cannot run anything from it, and the patchelf command would actually attempt to patch the binary with an interpreter which does not have the architecture expected by the binary.

**Comment by Havvy** [comment]
This seems to have stalled.

**Comment by bkchr** [comment]
Sorry :/ I did not forget the pull request! Too much to do at the moment.

**Comment by Havvy** [comment]
Still stalled;

**Comment by bkchr** [comment]
I finally find time to integrate the requested changes. I started with a complete rewrite of the pull request. 
I hope that it is better this time :)

**Comment by bkchr** [comment]
@nbp is this pull request still on your radar? :)

File: rust-overlay.nix [hunk 2]
```diff
@@ -39,27 +39,59 @@ let
...
       extensionNamesList;
 
-  getFetchUrl = pkgs: pkgname: stdenv: fetchurl:
+  hasTarget = pkgs: pkgname: target:
+    let
+      inherit (super.lib) hasAttrByPath;
+    in
+      (hasAttrByPath [ target ] pkgs.${pkgname}.target);
+
+  getTuples = pkgs: name: targets: builtins.map (target: { inherit name; inherit target; }) (builtins.filter (target: hasTarget pkgs name target) targets);
```
Comment (nbp) [inline rust-overlay.nix]: nit: { inherit name; inherit target; } --> { inherit name target; }
nit: add new lines.

File: rust-overlay.nix [hunk 3]
```diff
@@ -39,27 +39,59 @@ let
...
+      extensionNamesList = unique (builtins.map (pkg: pkg.pkg) extensions);
     in
       extensionNamesList;
 
-  getFetchUrl = pkgs: pkgname: stdenv: fetchurl:
+  hasTarget = pkgs: pkgname: target:
+    let
+      inherit (super.lib) hasAttrByPath;
+    in
+      (hasAttrByPath [ target ] pkgs.${pkgname}.target);
```
Comment (nbp) [inline rust-overlay.nix]: nit: there is a builtin operator for this operation now:
  pkgs ? ${pkgname}.target.${target}

File: rust-overlay.nix [hunk 4]
```diff
@@ -39,27 +39,59 @@ let
...
+
+  getTuples = pkgs: name: targets: builtins.map (target: { inherit name; inherit target; }) (builtins.filter (target: hasTarget pkgs name target) targets);
+
+  # Returns the { name, target } tuples for one package.
+  # If the package has components and these components support one of the targets, we add the components to the returning list with all supported targets.
+  # For the given pkgname, we only check the [ host "*" ] target list, because we don't want to install package for other architectures.
```
Comment (nbp) [inline rust-overlay.nix]: nit: (based on the current code)
In the manifest, a package might have different components which are bundled with it, as opposed as the extensions which can be added.  By default, a package will include the components for the same architecture, and offers them as extensions for other architectures.

This functions returns a list of { name, target } attribute sets, which includes the current system package, and all its components for the selected targets.
The list contains the package for the current host as well as the packages for components for all other targets

File: rust-overlay.nix [hunk 5]
```diff
@@ -68,8 +100,13 @@ let
           Select extensions from the following list:
           ${concatStringsSep "\n" availableExtensions}'';
       pkgsToInstall = [pkgname] ++ extensionsToInstall;
+      pkgsTuples = flatten(map (name: getTargetPkgTuples pkgs name targets stdenv) pkgsToInstall);
```
Comment (nbp) [inline rust-overlay.nix]: nit: add a space between "flatten" and "(map".
nit: based on the previous comment/code, `getTargetPkgTuples` will install extensions for the current host, but not for the targets, which would be the expected behavior.

Comment (bkchr) [inline rust-overlay.nix]: getTargetPkgTuples returns for the current host and for the targets?

**Comment by bkchr** [comment]
Updated :)

File: rust-overlay.nix [hunk 6]
```diff
@@ -68,8 +100,13 @@ let
           Select extensions from the following list:
           ${concatStringsSep "\n" availableExtensions}'';
       pkgsToInstall = [pkgname] ++ extensionsToInstall;
+      pkgsTuples = flatten (map (name: getTargetPkgTuples pkgs name targets stdenv) pkgsToInstall);
```
Comment (nbp) [inline rust-overlay.nix]: Given the following manifest:

```
[pkg.rust-analysis.target.aarch64-apple-ios]
available = true
hash = "977fcdc3e661797bdf25317ab7b629c655d636af4f3e550e1f2fc50a470a5be4"
url = "https://static.rust-lang.org/dist/2017-03-03/rust-analysis-nightly-aarch64-apple-ios.tar.gz"

[pkg.rust-analysis.target.i686-unknown-linux-gnu]
available = true
hash = "1efde22b019df35dbf8132a413fec68dc6cb858c83afd9109c15429963de34b7"
url = "https://static.rust-lang.org/dist/2017-03-03/rust-analysis-nightly-i686-unknown-linux-gnu.tar.gz"

[pkg.rust.target.x86_64-unknown-linux-gnu]
available = true
hash = "80353c4852088c182aa34ff5b30dd019106c7c85584a405ad793912cfabe1b0b"
url = "https://static.rust-lang.org/dist/2017-03-03/rust-nightly-x86_64-unknown-linux-gnu.tar.gz"

[[pkg.rust.target.x86_64-unknown-linux-gnu.components]]
pkg = "rustc"
target = "x86_64-unknown-linux-gnu"

[[pkg.rust.target.x86_64-unknown-linux-gnu.components]]
pkg = "rust-std"
target = "x86_64-unknown-linux-gnu"

[[pkg.rust.target.x86_64-unknown-linux-gnu.components]]
pkg = "cargo"
target = "x86_64-unknown-linux-gnu"

[[pkg.rust.target.x86_64-unknown-linux-gnu.extensions]]
pkg = "rust-docs"
target = "x86_64-unknown-linux-gnu"

[[pkg.rust.target.x86_64-unknown-linux-gnu.extensions]]
pkg = "rust-std"
target = "aarch64-apple-ios"

[[pkg.rust.target.x86_64-unknown-linux-gnu.extensions]]
pkg = "rust-analysis"
target = "aarch64-apple-ios"

[[pkg.rust.target.x86_64-unknown-linux-gnu.extensions]]
pkg = "rust-analysis"
target = "x86_64-unknown-linux-gnu"
```

and the following arguments:

```nix
{ extensions = [ "rust-analysis" ]; targets = [ "aarch64-apple-ios" ]; }
```

If I understand correctly, the current code will list the following packages:

```nix
[ { name = "rust"; target = "x86_64-unknown-linux-gnu"; }
  { name = "rust-analysis"; target = "x86_64-unknown-linux-gnu"; }
]
```

instead of

```nix
[ { name = "rust"; target = "x86_64-unknown-linux-gnu"; }
  { name = "rust-analysis"; target = "x86_64-unknown-linux-gnu"; }
  { name = "rust-analysis"; target = "aarch64-apple-ios"; }
]
```

Is that correct, or did I miss-interpret the intersection from `getTargetPkgTuples` ?

Comment (bkchr) [inline rust-overlay.nix]: Yeah that is correct, but why would you want to have rust-analysis for aarch64-apple-ios on an x86_64 machine?

Comment (bkchr) [inline rust-overlay.nix]: You have written by yourself "I think the current modification is slightly to ambitious and tries to address problem that we do probably should not care to support here (installing rust for arm, while running on x86_64 for example).". Wouldn't that be the case if we install the packages as you propose?

Comment (nbp) [inline rust-overlay.nix]: > why would you want to have rust-analysis for aarch64-apple-ios on an x86_64 machine?

I have no idea what rust-analysis is, but it is provided as an extension which is architecture dependent, such as rust-std.  If it is listed as an extension, there must be some interest to have it.

Comment (nbp) [inline rust-overlay.nix]: > Wouldn't that be the case if we install the packages as you propose?

I am not suggesting to install rust for arm, I am suggesting to install extensions which for different architecture if they are listed in the target list.  Setting `target = [ "aarch64-apple-ios" ];` should install all components/extensions for this target architecture as well, but only as part of derivations for the current architecture.

Basically, I meant that we should not install rust for aarch64-apple-ios, unless this is becomes an extension of another package.  The same way as we should not install rust-std for arm, unless this is an extension of rust.

Comment (bkchr) [inline rust-overlay.nix]: Ahh. Rust analysis contains informations about the libraries and is used by the rust language server for example to provide autocompletion etc. 

In my opinion, we should only install for the host architecture and the targets are for components. Because, if we install rust-analysis ios on x86 that does not give us anything or if we install cargo ios on x86. We can't not run these programs. But for components that target is important, because we want to have cross compilation.

So I would say, we should not install extensions for the target architectures, as they would be not usable on the host architecture.

**Comment by bkchr** [comment]
@nbp ping. Sorry if I'm annoying you.

**Comment by nbp** [comment]
@bkchr 
> @nbp ping.

Sorry, I meant to look at rustup, to see how they handle these cases.

One thing I recall is that components are changing, and some components gets integrated while other are moved as extension.  Knowing that, having a different installation when a component moves from being a component to being an extension would lead to having a non-working cross compiler.

**Comment by bkchr** [comment]
Yeah your are right, but for cross compilation we only need ruststd for the target platform installed. I hope that they does not remove that as component. What is your overall feeling for this pull request? Does I need to change something fundamentally? Or do we maybe stick to the current implementation and change if something changes in the future?

**Comment by nbp** [comment]
I think the next step is to install cross-architecture extensions, even if this does not seems to make sense for rust-analysis.

**Comment by bkchr** [comment]
But that means, if we have added "cargo" as an extension, we would install cargo compiled for a different architecture. I'm still not that convinced :D I could add an parameter e.g. "install_extensions_for_target_arch" and if that is set to true, we would install extensions for the target arch. What is your opinion on that?

**Comment by nbp** [comment]
@bkchr Having a `targetExtensions` list of extensions names sounds like a good approach which would solve these issues.

**Comment by bkchr** [comment]
Okay, I added the targetExtensions parameter.

File: rust-overlay.nix [hunk 7]
```diff
@@ -39,37 +39,85 @@ let
...
+      extensionsToInstall;
+
+  getSrcs = pkgs: pkgname: targets: extensions: targetExtensions: stdenv: fetchurl:
+    let
+      inherit (builtins) head map;
+      inherit (super.lib) flatten remove subtractLists unique;
+      targetExtensionsToInstall = checkMissingExtensions pkgs pkgname stdenv targetExtensions;
+      extensionsToInstall = unique ((checkMissingExtensions pkgs pkgname stdenv extensions) ++ targetExtensionsToInstall);
       pkgsToInstall = [pkgname] ++ extensionsToInstall;
+      pkgsTuples = flatten (map (name: getTargetPkgTuples pkgs name targets stdenv targetExtensionsToInstall) pkgsToInstall);
```
Comment (nbp) [inline rust-overlay.nix]: I do not understand this `map`, which calls for each `extensionsToInstall` the function to get the tuples with  `targetExtensionsToInstall`.  Logically, one of them seems redundant.

File: rust-overlay.nix [hunk 8]
```diff
@@ -39,37 +39,85 @@ let
...
+  # The list contains the package for the current host as well as the packages for components for all other targets
+  getTargetPkgTuples = pkgs: pkgname: targets: stdenv: targetExtensions:
+    let
+      inherit (builtins) elem;
+      inherit (super.lib) intersectLists;
+      pkg = pkgs.${pkgname};
+      components = getComponentsWithFixedPlatform pkgs pkgname stdenv;
+      extensions = getExtensions pkgs pkgname stdenv;
+      compExtIntersect = intersectLists components extensions;
+      pkgTargets = if elem pkgname targetExtensions then targets else [ "*" (hostTripleOf stdenv.system)];
```
Comment (nbp) [inline rust-overlay.nix]: This sounds really weird, and sounds like a work-around the comment on `getSrcs`.

File: rust-overlay.nix [hunk 9]
```diff
@@ -39,37 +39,85 @@ let
...
+  # In the manifest, a package might have different components which are bundled with it, as opposed as the extensions which can be added.
+  # By default, a package will include the components for the same architecture, and offers them as extensions for other architectures.
+  #
+  # This functions returns a list of { name, target } attribute sets, which includes the current system package, and all its components for the selected targets.
+  # The list contains the package for the current host as well as the packages for components for all other targets
+  getTargetPkgTuples = pkgs: pkgname: targets: stdenv: targetExtensions:
+    let
+      inherit (builtins) elem;
+      inherit (super.lib) intersectLists;
+      pkg = pkgs.${pkgname};
```
Comment (nbp) [inline rust-overlay.nix]: This attribute is no longer used.

File: rust-overlay.nix [hunk 10]
```diff
@@ -39,37 +39,85 @@ let
...
+      components = getComponentsWithFixedPlatform pkgs pkgname stdenv;
+      extensions = getExtensions pkgs pkgname stdenv;
+      compExtIntersect = intersectLists components extensions;
+      pkgTargets = if elem pkgname targetExtensions then targets else [ "*" (hostTripleOf stdenv.system)];
+      tuples = (getTuples pkgs pkgname pkgTargets) ++ (builtins.map (name: getTuples pkgs name targets) compExtIntersect);
```
Comment (nbp) [inline rust-overlay.nix]: why not doing:

```nix
let
  tuples =
     # List extensions which are not pre-installed components
     map (ext: getTuples pkgs ext targets) extensionSubComponents
     # List extensions which are pre-installed components excluding the current target.
     ++ map (ext: getTuples pkgs ext targetsExcludingThis) extensionAndComponents
     # Current package.
     ++ getTuples pkgs pkgname [ "*" (hostTripleOf stdenv.system)];
in
  tuples
```

**Comment by garbas** [comment]
@nbp @Havvy would it be possible to merge this so it doesn't get stalled?

**Comment by nbp** [comment]
@bkchr should I continue on what you already made, or you are going to continue this work later?

**Comment by bkchr** [comment]
Here we go again ;) Sorry for the long delay. I hope that we finally can merge this!

File: rust-overlay.nix [hunk 11]
```diff
@@ -39,37 +39,85 @@ let
...
-      pkgsToInstall = [pkgname] ++ extensionsToInstall;
     in
-      (builtins.map (pkg: (getFetchUrl pkgs pkg stdenv fetchurl)) pkgsToInstall);
+      extensionsToInstall;
+
+  getSrcs = pkgs: pkgname: targets: extensions: targetExtensions: stdenv: fetchurl:
+    let
+      inherit (builtins) head map;
+      inherit (super.lib) flatten remove subtractLists unique;
+      targetExtensionsToInstall = checkMissingExtensions pkgs pkgname stdenv targetExtensions;
```
Comment (nbp) [inline rust-overlay.nix]: follow-up: `targetExtensionsToInstall` is never used, and `checkMissingExtensions` is never called with the `extensions` list. Thus no error messages would be reported in case a non-existing extension is being listed in either list (`extensions` and `targetExtensions`).

File: rust-overlay.nix [hunk 12]
```diff
@@ -39,37 +39,85 @@ let
...
+      hostTargets = [ "*" (hostTripleOf stdenv.system)];
+      pkgTuples = flatten (getTargetPkgTuples pkgs pkgname hostTargets targets stdenv);
+      extensionTuples = flatten (map (name: getTargetPkgTuples pkgs name hostTargets targets stdenv) extensions);
+      targetExtensionTuples = flatten (map (name: getTargetPkgTuples pkgs name targets targets stdenv) targetExtensions);
```
Comment (nbp) [inline rust-overlay.nix]: These are way easier to read and understand than the previous iterations. 👍

**Comment by bkchr** [comment]
Thanks for merging @nbp :) I created #50 to address your last comments! I hope that are the last changes for this feature! :D
